### ISSUE TITLE
Unhandled Exception when network is unstable

### ISSUE THREAD

**Issue by KonH** [jweber/SlackBotNet#1]
Hello! I use your library in my project https://github.com/KonH/DotNetCoreBuildServer and ran into a problem, which I can resolve on my side (as I know it). I use your library there - https://github.com/KonH/DotNetCoreBuildServer/blob/master/Server/Integrations/SlackService.cs.

**Steps to reproduce:**
1. Initialize bot normally
2. Disable internet connection (simulate network problems)
3. Try to send any message
4. Observe results

**Actual result:**
```
Unhandled Exception: System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.WinHttpException: Не удается разрешить имя или адрес сервера
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext()
   --- End of inner exception stack trace ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Net.Http.HttpClient.<FinishSendAsync>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at System.Net.Http.HttpClient.<GetContentAsync>d__32`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at SlackBotNet.Drivers.SlackRtmDriver.<ConnectRtmAsync>d__6.MoveNext() in C:\projects\thirdparty\slackbotnet\src\SlackBotNet\Drivers\SlackRtmDriver.cs:line 84
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at SlackBotNet.Drivers.SlackRtmDriver.<>c__DisplayClass5_0.<<ConnectAsync>b__0>d.MoveNext() in C:\projects\thirdparty\slackbotnet\src\SlackBotNet\Drivers\SlackRtmDriver.cs:line 58
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```
When "Не удается разрешить имя или адрес сервера" is something like "Cannot resolve host name".

**Excepted result:**
Exceptions like this can be handled by library or provide some methods to handle it to library user code

Is something wrong with my integration? Or it can be fixed/improved inside library?
Waiting for your feedback, maybe I can help with it.

**Comment by jweber** [comment]
Thanks for the issue report. This is something I'd done some work on a while ago but never finished it. 

I just added the ability to hook into the point where a failure happens during the background operation that sends messages. Here's [the description](https://github.com/jweber/SlackBotNet#send-message-failure-handling) in the documenation.

It's published out to NuGet as [version 0.8.5](https://www.nuget.org/packages/SlackBotNet/0.8.5) which adds in diagnostic logging support as well as adding other changes to increase the reliability during network disruptions.

Let me know if this helps out with your error case or if it requires additional work. Thanks!

**Comment by KonH** [comment]
Great! Exactly what we need. Thanks for quick improvement.
